### PR TITLE
[ADD] feat(migration) : migration dspace 6.3 to dspace 8.3 

### PR DESCRIPTION
**Pull request by ayoub-chaalane**
## Checklist

- [x] Check the code style using the command `mvn clean && mvn install` on local desktop
- [ ] Run unitest for `dspace-uclouvain` module

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+     // CLASS ATTRIBUTES ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    public static final String USAGE_DESCRIPTION = "A command-line tool for importing versioning";
+    private ItemService itemService = ContentServiceFactory.getInstance().getItemService();
+    private VersioningService versioningService = VersionServiceFactory.getInstance().getVersionService();
+    private VersionHistoryService vsHistoryService = VersionServiceFactory.getInstance().getVersionHistoryService();
+    private EPersonService epersonService = EPersonServiceFactory.getInstance().getEPersonService();
+
+    public static void main(String[] argv) throws Exception {
+
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java]: COSMETIC: Try to limit non-relevant empty lines. It's ok to separate logical instruction blocks.

**Comment by Pourbaix** [comment]
You could squash your commits into a single one since there are not that much files in your PR.

File: dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        ImportVersioning versioning = new ImportVersioning();
+        CommandLine cl = versioning.validateCLIArgument(argv);
+
+        String fileString = cl.getOptionValue('f');
+
+        try {
+            versioning.run(fileString);
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java]: Could be merge into a single line 
```java
try {
    versionning.run(cl.getOptionValue('f'));
...
```

File: dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        } catch (Exception ex) {
+            logger.error(ex.getClass().getName() + " :: " + ex.getMessage(), ex);
+        }
+    }
+
+
+    private void run(String fileString) throws Exception {
+
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java]: remove non-relevant empty lines in all functions ;-)

File: dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        Context context = new Context();
+        context.turnOffAuthorisationSystem();
+
+        EPerson admin = epersonService.findByEmail(context, "admin@dspace.org");
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java]: Use a command-line argument to pass user_email (check `RabbitMQPackager.java`).
Hardcoding "admin@dspace.org" in this script will generate error when we will run this script on production server (admin@dspace.org is a TEST user, it will not be present on our production server)

File: dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        context.setCurrentUser(admin);
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java]: You need to test if `user != null`. If user is null, throw an `UserNotFoundException` (see RabbitMQPackager.java)

File: dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            versions.sort(Comparator.comparingInt(v -> v.get("version_number").asInt()));
+
+            for (JsonNode node : versions) {
+
+                int versionNumber = node.get("version_number").asInt();
+                String version_date = node.get("version_date").asText();
+                String summary = node.get("version_summary").asText();
+                String item_id = node.get("item_id").asText();
+                boolean in_archive = node.get("in_archive").asBoolean();
+                boolean discoverable = node.get("discoverable").asBoolean();
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java:129]: we use camelCase for variable naming/
* `version_date`--> `versionDate`
* `item_id` --> `itemID`
* ...

Pay attention to variable naming, this should be "itemUUID" instead of not "itemID"

File: dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+                itemChild.setArchived(in_archive);
+                itemChild.setDiscoverable(discoverable);
+                itemService.update(context, itemChild);
+
+                logger.info("Linked version " + versionNumber + " to item " + requestedUuid);
+            }
+        }
+
+        context.complete();
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java:155]: COSMETIC : this is non-sense here, but in further implementation, you should use `context.restoreAuthSystemState()` just after `context.complete()` to avoid security issues

File: dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    private Item findItemByIdentifierOer(Context context, String uuid)
+        throws SQLException, IOException, AuthorizeException {
+
+        Iterator<Item> it = itemService.findByMetadataField(
+                context, "dc", "identifier", "oer", uuid
+        );
+
+        return it.hasNext() ? it.next() : null;
```
Comment (zannkukai) [inline dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java]: * I **LOVE** ternary instruction but I **HATE** Java iterator.... It's a piece of shit to manipulate.
* Additionally, you should throws an exception if more than one item exists.
* "dc.identifier.oer" is ambigous. It should be "dc.identifier.legacyOER" or something like that

```java
    Iterator<Item> it = itemService.findByMetadataField(context, "dc", "identifier", "oer", uuid);
    if (!it.hasNext()) {
        return null;
    }
    Item first = it.next();
    if (it.hasNext()) {
        throw new SQLException("Multiple items found for identifier OER: " + uuid);
    }
    return first;
```

**Review by zannkukai — CHANGES_REQUESTED** [review]

File: dspace/config/crosswalks/mods-oer-submission.xsl [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+    <xsl:template match="//mods:mods">
+        <xsl:element name="dim:dim">
+            <xsl:apply-templates/>
+        </xsl:element>
+    </xsl:template>
+
+    <!-- GENRE -> dc.type ============================================= -->
+    <xsl:template match="/mods:mods/mods:genre">
+        <xsl:variable name="value">
+            <!-- todo : normaliser !!! -->
```
Comment (Pourbaix) [inline dspace/config/crosswalks/mods-oer-submission.xsl:57]: You forgot this TODO, or is it for a future change ?

File: dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    private void run(String epersonEmail, String fileString) throws Exception {
+        Context context = new Context();
+        context.turnOffAuthorisationSystem();
+
+        EPerson ePerson = Optional.ofNullable(epersonService.findByEmail(context, epersonEmail))
+            .orElseThrow(() -> new UserNotFoundException("Not found User : " + epersonEmail));
+        context.setCurrentUser(ePerson);
+
+        ObjectMapper mapper = new ObjectMapper();
+        JsonNode root = mapper.readTree(new FileInputStream(new File(fileString)));
```
Comment (Pourbaix) [inline dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java]: You can define your mapper in the json read statement like so:
```java
JsonNode root = new ObjectMapper().readTree(new FileInputStream(new File(fileString)));
```
This saves 1 line of code and a bit of memory.

File: dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+        logger.info("Starting version linking from JSON File: " + fileString);
+
+        Map<String, List<JsonNode>> grouped = new HashMap<>();
+        grouped = groupByRequestedUuid(root);
```
Comment (Pourbaix) [inline dspace-api/src/main/java/org/dspace/uclouvain/administer/ImportVersioning.java]: Same here you can reduce to 1 line:
```java
Map<String, List<JsonNode>> grouped = groupByRequestedUuid(root);
```

**Review by Pourbaix — COMMENTED** [review]
Some little changes

**Review by Pourbaix — APPROVED** [review]

**Review by zannkukai — APPROVED** [review]
